### PR TITLE
Add packageupdate

### LINKED ISSUE #21: Implement API PackageUpdate

### PR DESCRIPTION
**Pull request by leng20000**
I currently have some trouble implementing the PackageUpdate API, the error is about the lifetime parameter. What I'm trying to do is send in the Package to the API as an argument but it gives me the following error: 

image

https://github.com/lee3445/module-registry/blob/cdb700fef2b08f5073e22f9fcb12aaf61e0afc9d/src/api.rs#L30-L35

The error suggest to implement a FromRequest trait for the Package struct, but I'm still not sure how this works. Any thoughts on this? 

https://github.com/lee3445/module-registry/blob/cdb700fef2b08f5073e22f9fcb12aaf61e0afc9d/src/api/schema.rs#L43-L49

### CODE REVIEW

**Comment by starliteSeeker** [comment]
I think generally, you shouldn't need to implement `FromRequest` manually. Check out how I included `query` in `packages_list()`, but basically you will need `Deserialize` for the related structs, wrap the type in `Json<>` in the parameter, and include a `data = ...` in the `#[put...]` line.

Make sure to remove the unused imports.

Edit: `Serialize` is for sending out a json, and `Deserialize` for taking in a json, you'll want both.

**Comment by leng20000** [comment]
I fixed the error in the above commit. But I'm not sure how to test this API since the URL matches with the GET method instead, maybe we need restler for this?

File: src/api.rs [hunk 1]
```diff
@@ -27,6 +27,36 @@ pub fn test() -> &'static str {
...
+    if res.is_none() {
+        return (Status::NotFound, "Package does not exist.");
+    }
+
+    let db = res.unwrap();
+
+    //update the package if metadata matches
+    if (package.metadata.Name == db.name)
+        & (package.metadata.Version == db.ver)
+        & (package.metadata.ID == db.id)
```
Comment (starliteSeeker) [inline src/api.rs]: Even though `&` technically works, please use `&&` for my sanity.

Also, you'll want a `else` for when they don't match up, perhaps returning 404 or 400.

File: src/api.rs [hunk 2]
```diff
@@ -27,6 +27,36 @@ pub fn test() -> &'static str {
...
+    {
+        base64_to_zip(
+            package.data.Content.as_ref().unwrap().as_str(),
+            db.path.as_str(),
+        )
+        .await
+        .unwrap();
```
Comment (starliteSeeker) [inline src/api.rs]: Since the user might provide illegal base64, you'll have to handle the `Err` case.

**Comment by starliteSeeker** [comment]
Based on `@299` on piazza, I believe here's what should happen for the `Package` struct, `PackageRetrieve` and `PackageUpdate`:

`JSProgram` is always present, and at least one of `Content` or `URL` should be set. On `PackageRetrieve`, only `JSProgram` and `Content` should be set. On `PackageUpdate`, it should be either `JsProgram + Content` or `JsProgram + URL`.

Currently, your `PackageRetrieve` returns both `Content` and `URL`, the `PakcageData` struct has Option set on the wrong field, and `PackageUpdate` panics on `JsProgram + URL`

**Comment by leng20000** [comment]
I've made changes to `PackageData` struct so that `JSProgram` is a required field. So in the `package_retrieve` function I have put in the data for JSProgram as: 

https://github.com/lee3445/module-registry/blob/04549985e178baa96e8a7d5916811bb87a4c292b/src/api.rs#L179-L190

Does the `JSProgram` field here have the same value for all packages?

**Comment by starliteSeeker** [comment]
I think the JSProgram is one of the optional requirements and will be different for each package (if there is one), so since we're not implementing that, I'd say an empty string or the original `null` is fine.

Ignore this if you are working on it, but you haven't handled `URL` in `PackageUpdate` yet.

**Comment by leng20000** [comment]
Can you elaborate more on what `URL` field have to do in `package_update` function. Currently the function reads a `Package` struct input and extract the `Content` field to generate a zip file and update/replace the existing file. 

From what you are saying, package_update should handle `URL` field if the `Content` field is not set? So this means that package upload API have to be implemented first to get the content from the provided URL and update it? Am I understanding this correctly?

**Comment by starliteSeeker** [comment]
First question yes, second question I don't understand, but here's how I assumed it should be handled.

For both update and upload, if `Content` is not set, we use `URL` to first download the package, then save the zip at `db.path` (whereas if we have `Content`, we simply use `base64_to_zip` to write it to `db.path`). There will be repeated code in update and upload, so feel free to pull it out as a function (or not).

**Comment by leng20000** [comment]
Ok I understand it better now, but I was wondering if the other team implemented the download module function in Part 1? If so, I'll use that function in the package update API and it would be nice if you have reference to those function being called in the program.

**Comment by starliteSeeker** [comment]
From what I understand, no. They used GitHub API calls with octocrab only. The `ramp_up_time()` metrics in our project `src/metrics/github.rs` has cloning a repo, so you can check that out. [This](https://docs.github.com/en/repositories/working-with-files/using-files/downloading-source-code-archives) is also possible, but you might need to deal with the main/master branch names.

**Comment by leng20000** [comment]
I'm assuming this is the function you are referencing to: 

https://github.com/lee3445/ECE461_Team19_CLI/blob/555bf342ce30b51ca75a3da71dc092a38a9438b1/src/metrics/github.rs#L132-L160

But I'm not sure when did we import `git2` module in line `139`?

**Comment by starliteSeeker** [comment]
It's in `Cargo.toml`, if it's the "base" you don't need to `use git2`.

**Comment by starliteSeeker** [comment]
I have no way of telling if you're finished with your changes or not, so I'm just going to go ahead.

`git2` clones a repository and saves it in a folder, but we actually want it as a zip file. So, to use git2, you'll need to find a way to zip a file (I believe it's possible to use the `zip` crate, but I haven't tried it). Also, the URL provided by the user is a npm url like "https://www.npmjs.com/package/postcss", which is not a git repository.

The JSProgram is provided when user upload/updates modules. What should happen is that there should be a field in `Module` that holds the user-provided jsprogram, but since we're not planning to implement the functionality, I left it out when creating the struct. The `JsProgram` field in `PackageData` should be empty in package_retrieve (I said either null or empty string was fine, but now I think empty string is better)

**Comment by starliteSeeker** [comment]
Change the `cli` crate to expose functions that extract owner and repo name from url, and add a function `wget` that downloads a file using url. I added it in `cli` to avoid having to include reqwest again.

Handled the url case in `package_update`, and changed `JSProgram` to return empty string. I *think* I changed everything I mentioned above, but fell free to double check.

**Review by starliteSeeker — APPROVED** [review]

**Pull request merged**